### PR TITLE
feat: redirect to workspace after login with proper error handling

### PR DESCRIPTION
**Pull request by gentamura**
## Summary
- Add support for redirecting users to their originally requested workspace URL after login
- Prevent infinite redirect loops when authenticated users access workspaces they don't have permission for
- Improve type safety with proper type guards

## Changes
- Add `returnUrl` parameter to preserve original URL when redirecting to login
- Update middleware to include returnUrl in login redirect
- Support returnUrl in both OAuth and email/password login flows
- Add shared auth types (`AuthPageProps`, `AuthComponentProps`) to avoid duplicate type definitions
- Implement `isValidReturnUrl` type guard for better type safety and validation
- Handle 404 errors properly:
  - Unauthenticated users: Redirect to login with returnUrl (handled by middleware)
  - Authenticated users without access: Show 404 page with link to apps

## Implementation Details
- Validate returnUrl to prevent open redirect attacks (only relative URLs allowed)
- Use proper type guards instead of type assertions for FormData handling
- Clean code structure without need for `@ts-ignore` directives

## Test plan
1. **Unauthenticated user with valid workspace:**
   - Log out of the application
   - Access a workspace URL directly (e.g., `/workspaces/some-workspace-id`)
   - Verify you're redirected to login page with `returnUrl` parameter
   - Log in with email/password or OAuth
   - Verify you're redirected to the original workspace URL after successful login

2. **Authenticated user without workspace access:**
   - Log in as User A
   - Try to access User B's workspace URL
   - Verify you see a 404 page with "Go back to apps" link
   - Verify no infinite redirect loop occurs

3. **Edge cases:**
   - Test with invalid workspace IDs
   - Test with malformed returnUrl parameters
   - Verify OAuth flow preserves returnUrl correctly

___

### **PR Type**
Enhancement

___

### **Description**
- Add returnUrl parameter to preserve workspace URLs after login

- Implement type-safe validation to prevent open redirect attacks

- Support returnUrl in OAuth and email/password authentication flows

- Add global 404 page for unauthorized workspace access

___

### **Changes diagram**

```mermaid
flowchart LR
  A["User accesses workspace"] --> B["Middleware checks auth"]
  B --> C["Unauthenticated?"]
  C -->|Yes| D["Redirect to login with returnUrl"]
  D --> E["User logs in"]
  E --> F["Validate returnUrl"]
  F --> G["Redirect to original workspace"]
  C -->|No| H["Check workspace access"]
  H -->|No access| I["Show 404 page"]
  H -->|Has access| J["Show workspace"]
```

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>10 files</summary><table>
<tr>
  <td><strong>actions.ts</strong><dd><code>Add returnUrl support to OAuth providers</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/1265/files#diff-61d7135be22be4711eef8629730da8da2479354c19c02e7e73b3bc14af6527f0">+13/-7</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>index.ts</strong><dd><code>Export new returnUrl validation function</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/1265/files#diff-52455b389a38cb65c3550e0271b28e81a9db222e14fec7fd99c9bed3441a1219">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>validate-return-url.ts</strong><dd><code>Create type guard for returnUrl validation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/1265/files#diff-f7d9526914656fa5b1fc9f183bdfa11c8d29ccf955dd6dbd266175175cf9f1fb">+11/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>login.ts</strong><dd><code>Add returnUrl handling to email login</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/1265/files#diff-ad01a43bd06ccd096d01eb6e27d7f74d12d35716c2d31e3f548037fa89903ade">+8/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>types.ts</strong><dd><code>Define shared auth component types</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/1265/files#diff-9e4f60d5b15d47ccf07ded79e143108b1a1808d4807c12b76f09acee3048cee8">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>middleware.ts</strong><dd><code>Add returnUrl parameter to login redirects</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/1265/files#diff-7928989949ace78c38404a0fbbd358e48e544d028460d2f52066e9bc4ef3b9ba">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>oauth-providers.tsx</strong><dd><code>Pass returnUrl to OAuth forms</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/1265/files#diff-99bbaa52c4c38a169d0b8b11f34e13d7bb82497ad2174bc1219b69fef92ee333">+12/-2</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>login-form.tsx</strong><dd><code>Add returnUrl hidden input to form</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/1265/files#diff-817c5e2830b270cddda30aaec3c24ac102bf9425fcf6ff8be0732334a57e2470">+3/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>page.tsx</strong><dd><code>Extract returnUrl from search params</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/1265/files#diff-37e2dfb4577c27c93129c05b0bbc36e597ef70ac6400913e23f86e2579d108a7">+5/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>not-found.tsx</strong><dd><code>Create global 404 page component</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/1265/files#diff-6352ec79f2321552b57ecfc3f8455e7fb342ddd072f438c6bfecbe41fba47e1c">+15/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

## Summary by CodeRabbit

* **New Features**
  * Users are redirected back to their originally requested page after login via dynamic return URL support.
  * Added a custom 404 error page with a clear message and a link to return to the main apps page.

* **Enhancements**
  * OAuth and login forms now support an optional return URL to improve post-authentication navigation.
  * Return URLs are validated to ensure they are safe and confined within the application.

### CODE REVIEW

File: apps/studio.giselles.ai/app/(auth)/login/login.ts [hunk 1]
```diff
@@ -25,6 +26,9 @@ export async function login(
 			name: error.name,
 		};
 	}
-	redirect("/apps");
+
+	// Validate returnUrl to prevent open redirect attacks
+	const validReturnUrl = returnUrl?.startsWith("/") ? returnUrl : "/apps";
+	redirect(validReturnUrl);
 	return null;
```
[the remark this answers is not in the document]
Comment (gentamura) [reply apps/studio.giselles.ai/app/(auth)/login/login.ts:36]: In this case, I'm intentionally keeping the return null; statement after redirect() to make the control flow explicit and satisfy some TypeScript expectations. While redirect() throws internally and the return is technically unreachable, keeping it can improve clarity and avoid misleading type errors.

**Comment by gentamura** [comment]
@shige  @toyamarinyon Could you please review this code? :pray:

**Comment by gentamura** [comment]
> @shige @toyamarinyon Could you please review this code? 🙏

I'm checking the review feedback from the AI, so please hold on a moment. :pray:

File: apps/studio.giselles.ai/app/not-found.tsx [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+			<h1 className="text-4xl font-bold mb-4">404 - Page Not Found</h1>
+			<p className="text-gray-600 mb-8">
+				The page you are looking for does not exist.
+			</p>
+			<a href="/apps" className="text-blue-500 hover:underline">
+				Go back to apps
+			</a>
+		</div>
+	);
+}
```
[the remark this answers is not in the document]
Comment (gentamura) [reply apps/studio.giselles.ai/app/not-found.tsx]: fixed [23ccb2a](https://github.com/giselles-ai/giselle/pull/1265/commits/23ccb2a39b0bbf6f77a789ffd8d89e05257a05a9)

**Comment by gentamura** [comment]
> ### Bug: OAuth Flow Fails to Validate Return URL
> The OAuth authentication flow is vulnerable to open redirect attacks. The `returnUrl` parameter, extracted from `FormData`, is passed directly to `getAuthCallbackUrl` without validation. This contrasts with the email/password authentication flow, which properly validates `returnUrl` using `isValidReturnUrl()` to ensure it's an internal path (starts with `/`). This omission allows attackers to redirect users to arbitrary external sites after successful authentication.
> 
> `apps/studio.giselles.ai/app/(auth)/actions.ts#L8-L15`
> [Fix in Cursor](https://cursor.com/open?data=eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6ImJ1Z2JvdC12MSJ9.eyJ2ZXJzaW9uIjoxLCJ0eXBlIjoiQlVHQk9UX0ZJWF9JTl9DVVJTT1IiLCJkYXRhIjp7InJlZGlzS2V5IjoiYnVnYm90Ojk4YjIyODZkLTExNTQtNDkyNy04ZjM4LTE5YjQxZTRmMjFhNSIsImVuY3J5cHRpb25LZXkiOiJVLXFFNUJ0QmEwUnZHeTZFaEhabTRJSkxHLXVONUlFY29KYmYxQ1NyUXAwIiwiYnJhbmNoIjoiZmVhdC9yZWRpcmVjdC10by13b3Jrc3BhY2UtYWZ0ZXItbG9naW4ifSwiaWF0IjoxNzUxNTIyMDYzLCJleHAiOjE3NTIxMjY4NjN9.dvnvC8ntKLFiXeC9UeZ1Doi83MHIALblgKFZ2bAm-Y151Ru_ajKwOqpnZt6wp1xyVwOc_4OtzvMDYVKIfhYp4O37HYZwEDTWiFQx-34hyKtiJLoZrH87B43ZHExmyqgERRWJHUTifPViEs8cdrxaVaXoibzDMCLkVAwcN6RS_TRXWvLw82m6DDncj00gKRoY4vhSQGP1F8dnTVBlepm81EzBahLsOuONqlLWcprWp3RZs0edJwYQCcEVw9GXWv7jsIF7AZt6hkj4zcZHWuytUma5ktvFvOUHefMaCrTnNF51KbfPL3gLeNEMi8m7gA9H_htH3pBh3UITCxF6XzDGrQ)
> 
> _Was this report helpful? Give feedback by reacting with 👍 or 👎_

fixed [868af3e](https://github.com/giselles-ai/giselle/pull/1265/commits/868af3ed69189c530ec2d706d513d16e9017eb41)

**Comment by gentamura** [comment]
@shige @toyamarinyon Thanks for waiting. Could you please review this code? 🙏

**Review by toyamarinyon — APPROVED** [review]

**Comment by gentamura** [comment]
Thanks for reviewers 🚀
